README.md
# Überwachungstool für Fröling P2 / Lambdatronic S 3100 — ESP32-Adaptation

> **ESP32-Adaptation** von BeGrie — basierend auf dem ursprünglichen Raspberry-Pi-Projekt von [Daniel Höpfl (dhoepfl/Radiator)](https://github.com/dhoepfl/Radiator).

## Anpassung für ESP32 (BeGrie, Sept./Okt. 2022)

Im Zuge der Energiekrise 2022 und einem Pelletspreisanstieg von 200 auf 800 Euro je Tonne habe ich mein seit langem geplantes und immer wieder beiseite geschobenes Projekt (Heizung seit 2006 im Betrieb) der Heizungsüberwachung wieder aktiviert um Optimierungspotentiale zu erschließen.

Bei der Recherche nach Lösungen bin ich unter anderem auf dieses Projekt von Daniel Höpfl gestoßen und war von Analyse des P2/S3100-Kommunikationsprotokolls und der professionellen Umsetzung in Sourcecode begeistert. Hiermit huldige ich dem Meister, von dem ich aus seinem Projekt einiges lernen durfte! Vielen Dank dafür!

Als ESP32 Nutzer habe ich mangels RasPi Erfahrungen und fehlender Erläuterungen im Projekt (sorry) doch etwas gebraucht, um den gesamten Ablauf auf dem RasPi mittels C++, Autoconf/Automake, Python und Shellskripten in Kombination mit mysql, rrd, Vu, mailserver etc. zu verstehen. Nun ist alles klar, aber ich brauche so viele Features davon nicht und bleibe doch lieber in den mir bekannten ESP32-Gefilden ...

Die ESP32 Lösung besteht nur noch aus dem C++ Sourcecode, bei dem ich im wesentlichen folgendes geändert bzw. ergänzt habe:

- main.cpp neu für ESP32 mit Platformio (bei mir in VSCode) -> Platform Espressif32 und Arduino framework 3.x

- Serieller Input (in serial.cpp) über Serial2 realisiert (hardwareseitig wie beim RasPi über RS232-Schnittstellenumsetzer → bei mir SparkFun Transceiver Breakout - MAX3232 https://www.sparkfun.com/products/11189)

- Output fast komplett neu (output.h/.cpp, network.h/.cpp, files.h/.cpp) → nun "autark" (ohne externe Skripte o.ä) auf
  - Console/Serial,
  - Flash/LittleFS (wg. geringem Speicherplatz nur zum Testen sinnvoll),
  - SD Card (hardwareseitig jeder 3,3V SD Reader an den ESP32 anschliessbar). Aktuell muss die SD Card entnommen und am PC eingelesen werden.
  - Buzzer (aktiv) und Quit-Button zur akustischen Fehlermeldung im Keller
  - WiFi ...
    _ ... inkl. Konfiguration über console/Serial per USB-Anschluss → beim Booten den Quit-Button drücken
    _ ... zum Versand von MQTT-Nachrichten (Konfiguration wie bei WiFi) \* ... mit Webserver zum Datenabgriff von der SD-Card

- Die Konfiguration bzgl. Filesystem, Output und weiterem Verhalten erfolgt ausschließlich über die #defines in config.h, was bei Änderungen natürlich eine Neucompilierung erfordert (ich weiss: nicht schön - aber der kürzeste Weg). Dies sollte aber durch die Nutzung von Platformio über die hier enthaltene platformio.ini problemlos möglich sein ...

- Schaltplan zur Konfiguration im Sourcecode:
  ![Schaltplan](Schaltplan.png)

- Fotos zur Umsetzung:
  ![Foto Platine oben](Platine_Oben.jpg)
  ![Foto Platine unten](Platine_Unten.jpg)
  ![Foto Heizung](Heizung1.jpg)

Falls jemand die ESP32 Anpassung nutzt, bin ich für Rückmeldungen, Verbesserungen, Pull requests etc. offen ...

---

## Hintergrund: Originaltext von Daniel Höpfl ([dhoepfl/Radiator](https://github.com/dhoepfl/Radiator))

_Der folgende Abschnitt stammt aus dem ursprünglichen Raspberry-Pi-Projekt, auf dem diese ESP32-Adaptation basiert. Er enthält die detaillierte Protokollanalyse der Fröling P2 / Lambdatronic S 3100 Schnittstelle._

## Vorgeschichte

Fast alle Geräte in unserem Haushalt haben einen eingebauten Signalgeber, um uns zu informieren, wenn sie etwas von uns brauchen: Die Spülmaschine piept, wenn sie fertig ist, die Mikrowelle hat eine laute Klingel, die Waschmaschine ebenso, wie der Trockner. Das einzige Gerät, das keinen Ton von sich gibt, wenn es eine Störung erkennt, ist die Fröling P2 Pellet-Heizung im Keller. Es ist nicht so, dass die Heizung keine Störungen hätte oder diese nicht erkennt, nein: Wenn zum Beispiel ein zu großes Pellet den Saugschlauch blockiert, dann erkennt die Heizung das und ihre Status-LED blinkt nicht länger grün sondern rot. Dass wir das nicht bemerken, weil die Heizung hinter verschlossener Türe im Heizungskeller steht, daran hat bei Fröling wohl niemand gedacht.

Mein erster Ansatz war, eine kleine Überwachung zu bauen, die die Farbe der LED erkennt und darauf basierend einen Alarm auslöst. Ich bin Softwerker, kein Hardware-Profi. An der Erkennung der Farbe (zu kleinem Preis) bin ich gescheitert. Noch während ich daran gearbeitet habe, bin ich darauf gestoßen, dass die P2 bzw. die darin verbaute Lambdatronic S 3100 eine serielle Schnittstelle hat, über die sie praktisch voll bedienbar ist. Da ich zunächst eine Lösung wollte, für die ich mich nicht direkt mit der Heizung verbinden muss (Gewährleistung, Diskussionen mit dem Service, ...), habe ich diese Möglichkeit zunächst nicht weiter verfolgt.

Nachdem wir in den letzten Wochen mehrmals (Pellets-Lager wird leer) durch kaltes Wasser auf die Farbe der LED hingewiesen wurden, habe ich das Projekt wiederbelebt. Dieses Mal ist mir der Mitarbeiter der Wartungsfirma egal.

## Protokollanalyse

Das Protokoll der S 3100 ist halbwegs gut analysiert. Die beste Quelle ist [dieser Thread bei mikrocontroller.net](http://www.mikrocontroller.net/topic/319990 "Fröling Heizungssteuerung auslesen"). Auch wertvoll ist [dieser Eintrag auf rigert.com](http://rigert.com/wiki-wiki/index.php?title=Fr%C3%B6ling). Die einzige Quelle, die „M3“ erwähnt ist [das hier verlinkte Archiv eines Foren-City-Threads](http://www.labviewforum.de/Thread-Mit-RS232-Daten-lesen-und-schreiben?pid=159201#pid159201).

Die drei Quellen scheinen größtenteils voneinander abzuschreiben (Forencity ist dabei der älteste, gefolgt von mikrocontroller und dann rigert.com), allen gemein ist, dass meiner Ansicht nach keiner der dort Mitschreibenden das Protokoll verstanden hat. Da wird davon gesprochen, dass 32 Mal ein bestimmter Befehl gesendet werden müsste, um die Kommunikation am Laufen zu halten, oder dass man abwechselnd „M1“ und „M2“ abfragen müsste. Ich halte das für Blödsinn.

Nach meinen Analysen sieht das Protokoll wie folgt aus:

### Datenpakete

Das Protokoll basiert auf Datenpaketen mit folgender Struktur:

| Bytes | Beschreibung |
| ----- | ------------ |
| 2     | Befehl       |
| 1     | Länge (n)    |
| n     | Daten        |
| 2     | Prüfsumme    |

Dabei gilt:

- Für eine Liste der bekannten Befehle siehe unten. Soweit ich sehen konnte, beginnen Befehle, die von der Heizung gesendet werden, mit „M“, Befehle, die vom Endgerät ausgehen, beginnen mit „R“. (Achtung: Das gilt nicht für Bestätigungen)
- Die Länge zählt nur die Datenbytes, Header und Prüfsumme bleiben unberücksichtigt.
- Die Prüfsumme ist eine einfache, vorzeichenlos berechnete, 16 Bit breite Summe über Befehl, Länge und Daten.

## Ablauf

Hier widerspreche ich den oben verlinkten Quellen: Meiner Meinung nach sieht das Protokoll vor, dass jedes Datenpaket mit einer Bestätigung quittiert wird. Dabei folgt das Bestätigungspaket folgendem Schema:

| Bytes | Beschreibung                                       |
| ----- | -------------------------------------------------- |
| 2     | Befehl, der quittiert wird                         |
| 1     | Länge (immer 1)                                    |
| 1     | 0x01 für Bestätigung (ACK), 0x00 für Fehler (NACK) |
| 2     | Prüfsumme                                          |

Der Ablauf im Programm kann also wie folgt aussehen:

1.  Benutze einen Puffer für mindestens 260 Bytes (2 Bytes Befehl, 1 Byte Länge, max. 255 Bytes Daten, 2 Bytes Prüfsumme).
2.  5 Bytes von der seriellen Schnittstelle lesen (2 Byte Befehl, 1 Byte Länge, 2 Bytes Prüfsumme)
3.  Wenn Länge > 0: Entsprechend weitere Bytes lesen.
4.  Die Prüfsumme prüfen.
    - Ist sie ungültig: NACK senden, weiter bei 1.
    - Ist sie gültig: ACK senden, weiter bei 5.
5.  Bestätigungen behandeln:
    - Wurde der entsprechende Befehl gesendet und ist die Länge 1 und ist der Wert 0x01? Dann ist der Befehl quittiert, weiter bei 1.
    - Wurde der entspechende Befehl gesendet und ist die Länge 1 und ist der Wert 0x00? Dann Befehl erneut senden, weiter bei 1.
    - Wurde der entsprechende Befehl nicht gesendet oder ist die Länge nicht 1, dann weiter bei 6.
6.  Befehl behandeln.

## Bekannte Befehle

Hier die Beschreibung der Befehle, so wie ich sie zusammengetragen und verstanden habe. Das unterscheidet sich zum Beispiel bei „M1“ von dem, was man in den Foren findet, dort wird immer von einer festen Reihenfolge ausgegangen. Für eine bestimmte Heizung stimmt das vermutlich auch, solange nichts neues angeschlossen wird.

Alle Texte, die übertragen werden, sind übrigens in „Codepage 850“ kodiert. Ein (Mapping zu Unicode findet sich z.B. bei der Wikipedia)[https://de.wikipedia.org/wiki/Codepage_850].

### Login („Ra“)

| Bytes | Wert      | Beschreibung                                                        |
| ----- | --------- | ------------------------------------------------------------------- |
| 2     | 0x52 0x61 | Login-Befehl                                                        |
| 1     | 0x03      | Länge 3                                                             |
| 1     | 0x00      | unbekannt                                                           |
| 2     | 0x.. 0x.. | Benutzer-Kennung (-7: 0xff 0xf9 für Service, 1/0x00 0x01 für Kunde) |
| 2     | 0x.. 0x.. | Prüfsumme                                                           |

Die Heizung antwortet darauf mit einem ACK (siehe Ablauf) und sendet daraufhin die Befehlsblöcke MA, MB, MC, MD, ME, MF, MG, MK, ML, MM, MW, MS, MT und MU, jeweils beendet durch entsprechende MZ Befehle. Anschließend beginnt die Heizung regelmäßig M2 zu senden.

### Status („Rb“)

| Bytes | Wert           | Beschreibung  |
| ----- | -------------- | ------------- |
| 2     | 0x52 0x62      | Status-Befehl |
| 1     | 0x03           | Länge 3       |
| 3     | 0x00 0x00 0x00 | unbekannt     |
| 2     | 0x.. 0x..      | Prüfsumme     |

Die Heizung antwortet darauf mit einem ACK (siehe Ablauf) und sendet ab sofort regelmäßig die Messwerte (M1).

### Parameter ändern („RI“)

Ändert einen der Werte der Heizung. Die verfügbaren Werte werden in „ME“ beschrieben.

| Bytes | Wert      | Beschreibung |
| ----- | --------- | ------------ |
| 2     | 0x52 0x49 | Befehl       |
| 1     | 0x04      | Länge        |
| 2     | 0x.. 0x.. | Parameter-ID |
| 2     | 0x.. 0x.. | Neuer Wert   |
| 2     | 0x.. 0x.. | Prüfsumme    |

Die Heizung antwortet darauf mit einem ACK/NACK (siehe Ablauf), anschließend kommt der neue Wert über „MI“.

### Messwerte („M1“)

Auch hier muss ich den Angaben aus den Foren widersprechen: Der Inhalt von M1 ist nicht fest vorgegeben, es hängt davon ab, was wie die Heizung installiert wurde. Welche Werte geliefert werden und wie diese zu interpretieren sind, kann man aus der „MA“- bzw. „MC“ Antworten ableiten, die man beim Login bekommt. Für meine Heizung sind das die folgenden Werte, für andere Heizungen können es andere Werte (oder Reihenfolgen) sein.

| Bytes | Wert      | Beschreibung           | Faktor | Einheit |
| ----- | --------- | ---------------------- | :----: | :-----: |
| 2     | 0x4D 0x31 | Status-Befehl          |        |         |
| 1     | 0x..      | Länge                  |        |         |
| 2     | 0x.. 0x.. | 1. Displayzeile        |        |         |
| 2     | 0x.. 0x.. | 2. Displayzeile        |        |         |
| 2     | 0x.. 0x.. | Zustand                |   1    |   -/-   |
| 2     | 0x.. 0x.. | Rost                   |   1    |   -/-   |
| 2     | 0x.. 0x.. | Kessel-Temperatur      |   2    |   °C    |
| 2     | 0x.. 0x.. | Abgas-Temperatur       |   1    |   °C    |
| 2     | 0x.. 0x.. | Abgas Schwellwert      |   1    |   °C    |
| 2     | 0x.. 0x.. | KesselStellGr          |   1    |    %    |
| 2     | 0x.. 0x.. | Saugzug                |   1    |    %    |
| 2     | 0x.. 0x.. | Zuluft-Gebläse         |   1    |    %    |
| 2     | 0x.. 0x.. | Einschub               |   1    |    %    |
| 2     | 0x.. 0x.. | Rest-O2                |  100   |    %    |
| 2     | 0x.. 0x.. | O2-Regler              |  100   |    %    |
| 2     | 0x.. 0x.. | Füllstand              |  207   |    %    |
| 2     | 0x.. 0x.. | Feuerraum-Temperatur   |   2    |   °C    |
| 2     | 0x.. 0x.. | Puffer oben            |   2    |   °C    |
| 2     | 0x.. 0x.. | Puffer unten           |   2    |   °C    |
| 2     | 0x.. 0x.. | Puffer Pu              |   1    |    %    |
| 2     | 0x.. 0x.. | Boiler                 |   2    |   °C    |
| 2     | 0x.. 0x.. | Außentemperatur        |   2    |   °C    |
| 2     | 0x.. 0x.. | Vorlauf 1 Schwellwert  |   2    |   °C    |
| 2     | 0x.. 0x.. | Vorlauf 1              |   2    |   °C    |
| 2     | 0x.. 0x.. | Vorlauf 2 Schwellwert  |   2    |   °C    |
| 2     | 0x.. 0x.. | Vorlauf 2              |   2    |   °C    |
| 2     | 0x.. 0x.. | KTY6_H2                |   2    |   °C    |
| 2     | 0x.. 0x.. | KTY7_H2                |   2    |   °C    |
| 2     | 0x.. 0x.. | Brennerstarts          |   1    |   -/-   |
| 2     | 0x.. 0x.. | Betriebsstunden        |   1    |    h    |
| 2     | 0x.. 0x.. | Board-Temperatur       |   1    |   °C    |
| 2     | 0x.. 0x.. | Kessel Soll-Temperatur |   2    |   °C    |
| 2     | 0x.. 0x.. | Prüfsumme              |        |         |

### Zeit („M2“)

| Bytes | Wert      | Beschreibung                           |
| ----- | --------- | -------------------------------------- |
| 2     | 0x4D 0x32 | Befehl                                 |
| 1     | 0x07      | Länge                                  |
| 1     | 0x..      | Sekunden                               |
| 1     | 0x..      | Minuten                                |
| 1     | 0x..      | Stunden                                |
| 1     | 0x..      | Tag                                    |
| 1     | 0x..      | Monat                                  |
| 1     | 0x..      | Wochentag (1: Montag, ..., 7: Sonntag) |
| 1     | 0x..      | Jahr                                   |
| 2     | 0x.. 0x.. | Prüfsumme                              |

Alle Werte sind BCD, d.h. die oberen 4 Bit beschreiben die Zehner-Stelle, die unteren 4 Bit die Einer-Stelle.

### Fehlermeldung („M3“)

Wird gesendet, wenn ein Fehler auftritt. Vermutlich die interessanteste Meldung überhaupt, jedenfalls für meinen Anwendungszweck!

| Bytes | Wert      | Beschreibung                           |
| ----- | --------- | -------------------------------------- |
| 2     | 0x4D 0x33 | Befehl                                 |
| 1     | 0x0a      | Länge                                  |
| 1     | 0x..      | Fehlertext-ID                          |
| 1     | 0x..      | Unbekannt                              |
| 1     | 0x..      | Unbekannt                              |
| 1     | 0x..      | Sekunden                               |
| 1     | 0x..      | Minuten                                |
| 1     | 0x..      | Stunden                                |
| 1     | 0x..      | Tag                                    |
| 1     | 0x..      | Monat                                  |
| 1     | 0x..      | Wochentag (1: Montag, ..., 7: Sonntag) |
| 1     | 0x..      | Jahr                                   |
| 2     | 0x.. 0x.. | Prüfsumme                              |

Der Inhalt entspricht dem von „MU“, die Fehlertext-ID ist analog denen in „MT“.

### Bezeichnung der Parameter in „M1“ („MA“)

Diese Nachrichten müssen ausgewertet werden, um den Inhalt von „M1“ auswerten zu können. Die Werte in „M1“ sind jeweils 2 Byte lang und werden in genau der Reihenfolge gesendet, in der auch die „MA“ Nachrichten gesendet wurden.

| Bytes | Wert      | Beschreibung                         |
| ----- | --------- | ------------------------------------ |
| 2     | 0x4D 0x41 | Befehl                               |
| 1     | 0x..      | Länge                                |
| 1     | 0x..      | 'S': String, 'I': Messwerte          |
| 2     | 0x.. 0x.. | 'S': Typ in „MB“, 'I': Index in „MC“ |
| 2     | 0x.. 0x.. | Unbekannt                            |
| n     | ...       | Parameterbezeichnung                 |
| 2     | 0x.. 0x.. | Prüfsumme                            |

### Displaytexte („MB“)

Enthält die Texte, die in „MA“ referenziert werden. Bisher bekannt sind die beiden Typen 0 und 1, welche in der ersten bzw. zweiten Displayzeile angezeigt werden.

| Bytes | Wert      | Beschreibung                                        |
| ----- | --------- | --------------------------------------------------- |
| 2     | 0x4D 0x42 | Befehl                                              |
| 1     | 0x..      | Länge                                               |
| 2     | 0x.. 0x.. | Typ (0: obere Displayzeile, 1: untere Displayzeile) |
| 2     | 0x.. 0x.. | Index der Nachricht                                 |
| n     | Text      | Nachrichtentext                                     |
| 2     | 0x.. 0x.. | Prüfsumme                                           |

### Datenformat der Werte in „M1“ („MC“)

Beschreibt, wie die Daten in „M1“ formatiert werden müssen. Da es bei mir deutlich mehr „MC“-Datensätze, als für „MA“ gibt, gehe ich davon aus, dass die übrigen Einträge anderweitig referenziert werden. Wo habe ich noch nicht herausgefunden.

| Bytes | Wert      | Beschreibung                                       |

NEWS
Raspirator
==========

February 2016: Initial release

September 2022: Adaption for ESP32 - standalone


src/main.cpp
/******************************************************************************
 * Name:		main.cpp
 * @brief   Überwachungstool für Fröling P2 / Lambdatronic S 3100
 *          -> hier: für ESP32
 *          -> Output auf console, flash-filesystem oder SD-card oder/und MQTT
 *          -> rudimentärer Zugriff per WiFi Async Server
 *          -> IDE: platformio
 * Created:	01.09.2022 (ESP32-adaption)
 * Author:	Original (for Raspberry Pi): Daniel Höpfl
 *          -> https://github.com/dhoepfl/Radiator
 *          Adaption to ESP32: Bernd Griesbach
 *          -> https://github.com/begrie/Radiator_Froeling_S3100_Hoepfl/tree/ESP32
 ******************************************************************************/
#include <Arduino.h>
#include <LittleFS.h>
#include <SPIFFS.h>
#include "FS_Filehelper.h"

#include "debug.h"
#include "device.h"
#include "surveillance.h"
#include "output.h"

/*********************
 *      DEFINES
 *********************/
#define D_DEBUG_LEVEL 3                       // 0 - No logging
                                              // 1 - Errors
                                              // 2 - Warnings
                                              // 3 - Infos
                                              // 4 - Debug
                                              // 5 - Trace
#define REDIRECT_STD_ERR_TO_SYSLOG_FILE false // debug output to file instead to console/Serial
                                              // -> watch the needed filespace:
                                              //    makes only sense at long term operation with D_DEBUG_LEVEL 1 - Errors

#define S_SERIAL_TO_RADIATOR "Serial2" // Serial device for connection to radiator device P2/S3100
// #define SERIAL_TO_RADIATOR_RX 16 // ESP32-standard: 16 -> needs to be changed in serial.cpp if neccessary
// #define SERIAL_TO_RADIATOR_TX 17 // ESP32-standard: 17 -> needs to be changed in serial.cpp if neccessary
#define T_TIMEOUT_BETWEEN_TRANSFERS_MS 5000 // Maximum time allowed between transfers from P2/S3100 radiator device
                                            // until connection is considered stale.

// Where to write the received values to:
#define OUTPUT_TO_CONSOLE false
#define OUTPUT_TO_MQTT false
#define OUTPUT_TO_FILE true
#define FILESYSTEM_TO_USE LittleFS           // LittleFS (or SPIFFS) or SD
#define DATA_DIRECTORY "/P2_Logging"         // directory for radiator data -> e.g. "/logging" or "" for root dir of used filesystem
#define MIN_FREE_KILOBYTES_ON_FILESYSTEM 300 // if limit is undercut -> oldest logfiles are deleted

// !!! SEE MORE #defines in output.h. FOR CONTROLLING THE OUTPUT BEHAVIOUR !!!

/********************************************
 * FORWARD DECLARATIONS
 ********************************************/
void xTask_Watchdog_and_Maintenance(void *parameter);
std::string initFilesystem(std::string dataDirectory);
void checkFreeSpaceOnFilesystem();
void writeSyslogfile();

/********************************************
 * GLOBAL DEFINITIONS
 ********************************************/
std::string pathnameToDataDirectory; // with mountpoint/mountdir of filesystem
bool outputToConsole = OUTPUT_TO_CONSOLE;
bool outputToMQTT = OUTPUT_TO_MQTT;

#if REDIRECT_STD_ERR_TO_SYSLOG_FILE
std::ofstream syslogFileStream;
std::string syslogFileName = "/littlefs/syslog1.log";
#endif

/*********************************************************************
 * @brief 	Setup to start everything
 * @param 	void
 * @return 	void
 *********************************************************************/
void setup()
{
  Serial.begin(115200);
  delay(1000); // give serial monitor time to connect

  debug_level = D_DEBUG_LEVEL;
  std::string dataDirectory = DATA_DIRECTORY; // directory for data from radiator -> e.g. "/data" or "" for root dir
                                              // all directories in path must exist before instantiation of OutputHandler
                                              // (only one level is created automatically in initFilesystem())

  if (!OUTPUT_TO_FILE)
    pathnameToDataDirectory = ""; // empty pathname indicates "no file output" for constuctor of OutputHandler
  else
  {
    pathnameToDataDirectory = initFilesystem(dataDirectory); // LittleFS (or SPIFFS or SD), directory for data
                                                             // patnameToFIles is completed with mountpoint of the filesystem
    if (pathnameToDataDirectory.empty())
    {
      std::cout << "ERROR MOUNTING FILESYSTEM !!! -> Instead: data output to console (std::cout)" << std::endl;
      outputToConsole = true;
    }
    else
    {
#if REDIRECT_STD_ERR_TO_SYSLOG_FILE
      // Ausgaben auf Logfile umleiten -> geht so nur mit regelmäßigem close und re-open des streams -> BUG bei LittleFS??
      LOG_info << millis() << " ms: !!! ALL DEBUG OUTPUT TO std::err IS NOW REDIRECTED TO FILE " << syslogFileName << " !!!\n"
               << "  \t !!! -> NO OUTPUT TO CONSOLE ANY MORE !!!" << std::endl;
      syslogFileStream.open(syslogFileName, std::ofstream::out | std::ofstream::app);
      if (syslogFileStream && syslogFileStream.is_open() && syslogFileStream.good())
      {
        //  syslogFilestream.rdbuf()->pubsetbuf(0, 0);  //sollte ohne buffer direkt schreiben -> kein Wirkung bei littlefs
        syslogFileStream << "Syslogfile " << syslogFileName << " opened successfull " << std::endl;
        std::cerr.rdbuf(syslogFileStream.rdbuf()); // redirect by assigning buffer of file stream
                                                   // std::cerr.tie(&syslogFilestream);
        syslogFileStream << "Redirect output from std::err to syslogfile " << std::endl;
      }
      else
      {
        LOG_info << "ERROR opening syslog file " << syslogFileName << std::endl;
      }
      LOG_info << "NACH       syslogFilestream.open('/littlefs/syslog.log');" << std::endl;
#endif

      TaskHandle_t Handle_xTask_Watchdog_and_Maintenance;

      // Create RTOS task
      BaseType_t _Result = xTaskCreatePinnedToCore(
          xTask_Watchdog_and_Maintenance,         // Task function
          "xTask_Watchdog_and_Maintenance",       // String with name of task
          4096,                                   // Stack size in bytes
          NULL,                                   // Parameter passed as input of the task
          uxTaskPriorityGet(NULL),                // Priority of the task: higher values -> higher priority
                                                  // with uxTaskPriorityGet(NULL)-> same priority as current task
          &Handle_xTask_Watchdog_and_Maintenance, // Task handle (Typ: TaskHandle_t)
          1);                                     // Core 0 or 1 (Arduino code by default on Core 1)

      if (_Result != pdPASS)
      {
        throw std::runtime_error("Start_Watchdog_and_Maintenance_Task: Error creating xTask");
      }
    }
  }
}

/*********************************************************************
 * @brief 	Loop function
 * @param 	void
 * @return 	void
 *********************************************************************/
void loop()
{
  try
  {
    // for (;;)
    // {
    LOG_trace << "Starting main loop" << std::endl;

    radiator::OutputHandler outHandler(pathnameToDataDirectory, // with empty pathnameToDataDirectory -> no output to file
                                       outputToConsole,
                                       outputToMQTT);
    radiator::Surveillance surveillance(S_SERIAL_TO_RADIATOR, T_TIMEOUT_BETWEEN_TRANSFERS_MS, outHandler);
    surveillance.main_loop();

    LOG_info << "Main loop ended, restarting in 10s" << std::endl;
    sleep(10);

    // }
  }
  catch (const char *&error)
  {
    LOG_fatal << "Failed to open serial device or filesystem -> restarting ESP32 in 10s" << std::endl;
    sleep(10);
    esp_restart();
  }
}

/*********************************************************************************************
 *@brief   task for cyclic actions like filesystem space check
 *         and watchdog regarding possible malfunctions -> still TODO
 *@param   pointer to parameter for the task -> nothing here!
 *@return  void
 *********************************************************************************************/
void xTask_Watchdog_and_Maintenance(void *parameter)
{
  vTaskDelay(pdMS_TO_TICKS(10000)); // wait 10 sec. at startup

  const auto intervallForFilesystemCheckSec = 6 * 60 * 60; // execute all 6 hours
  ulong nextFilesystemCheck = 0;

#if REDIRECT_STD_ERR_TO_SYSLOG_FILE
  const auto intervallForWriteSyslogfileSec = 15; // all 15 minutes
  // const auto intervallForWriteSyslogfileSec = 15 * 60; // all 15 minutes
  ulong nextWriteSyslogfile = 0;
#else
  const auto intervallForWriteSyslogfileSec = INT_MAX;
#endif

  const auto watchDogIntervallSec = std::min(intervallForFilesystemCheckSec, intervallForWriteSyslogfileSec);

  while (true) // the xTask runs "endless"
  {
    if (millis() > nextFilesystemCheck)
    {
      checkFreeSpaceOnFilesystem();
      nextFilesystemCheck = millis() + intervallForFilesystemCheckSec * 1000;
    }

#if REDIRECT_STD_ERR_TO_SYSLOG_FILE
    if (millis() > nextWriteSyslogfile)
    {
      writeSyslogfile();
      nextWriteSyslogfile = millis() + intervallForWriteSyslogfileSec * 1000;
    }
#endif

    vTaskDelay(watchDogIntervallSec * 1000 / portTICK_PERIOD_MS); // execute all xx seconds
  }
  assert(false); //-> we must never get here -> if we do -> restart
}

/*********************************************************************
 * @brief 	Initialization of filesystem for use with ostream
 * @param 	Filesystem to use: LittleFS (or SPIFFS) or SD
 * @param   Directory for data received from radiator e.g. "/data" or "" for root dir
 * @return 	Pathname to data directory with mountpoint/mountdir like "/littlefs/data"
 *********************************************************************/
std::string initFilesystem(std::string dataDirectory)
{
  std::string basepath;
  std::string fsName;

  if ((void *)&FILESYSTEM_TO_USE == (void *)&LittleFS)
  {
    fsName = "LittleFS";
    basepath = "/littlefs"; // standard mountpoint(mountpath)
  }
  else if ((void *)&FILESYSTEM_TO_USE == (void *)&SD)
  {
    fsName = "SD";
    basepath = "/sd"; // standard mountpoint(mountpath)
  }
  else if ((void *)&FILESYSTEM_TO_USE == (void *)&SPIFFS)
  {
    fsName = "SPIFFS";
    basepath = "/spiffs"; // standard mountpoint(mountpath)
  }

  LOG_info << "\nSelected filesystem for output to file: " << fsName << std::endl;

  if (!FILESYSTEM_TO_USE.begin(false))
  {
    LOG_info << fsName << " mount failed -> Try to format file system for " << fsName << " -> please wait ..." << std::endl;
    if (!FILESYSTEM_TO_USE.begin(true)) // open with format filesystem
    {
      LOG_error << fsName << " formatting not possible -> NO LOCAL DATA STORAGE AVAILABLE" << std::endl;
      return "";
    }
    LOG_info << "Formatting of filesystem for " << fsName << " successfull" << std::endl;
  }

  // FS_Filehelper::deleteDirectory("/");
  // FILESYSTEM_TO_USE.format();

  LOG_info << "Filesystem   " << fsName << "   was successfull mounted at basepath:  " << basepath
           << "  -> Available space " << ((FILESYSTEM_TO_USE.totalBytes() - FILESYSTEM_TO_USE.usedBytes()) / 1024)
           << " kB from total " << (FILESYSTEM_TO_USE.totalBytes() / 1024) << " kB" << std::endl;

  if (!dataDirectory.empty())
  {
    if (dataDirectory.front() != '/')
      dataDirectory = '/' + dataDirectory;

    if (FILESYSTEM_TO_USE.exists((dataDirectory).c_str())) // it seems that LittleFS creates a new dir on its own if it did not exist
    {
      LOG_info << "Directory " << dataDirectory << " already exists" << std::endl;
    }
    else
    {
      LOG_info << "dataDirectory " << dataDirectory << " did not exist -> try to create it ..." << std::endl;
      if (FILESYSTEM_TO_USE.mkdir(dataDirectory.c_str()))
      {
        LOG_info << "... " << dataDirectory << " was successfully created" << std::endl;
      }
      else
      {
        LOG_error << "... " << dataDirectory << " CAN NOT BE CREATED" << std::endl;
        dataDirectory = "";
      }
    }
  }

  basepath += dataDirectory;
  LOG_info << "Whole output path is  " << basepath << std::endl;

  FS_Filehelper::listDir("/", 255, true, FILESYSTEM_TO_USE);

  return basepath;
}

/*********************************************************************
 * @brief 	Initialization of filesystem for use with ostream
 * @param 	Filesystem to use: LittleFS or SPIFFS or SD
 * @param   Directory for data received from radiator e.g. "/data" or "" for root dir
 * @return 	Pathname to data directory with mountpoint/mountdir like "/littlefs/data"
 *********************************************************************/
void checkFreeSpaceOnFilesystem()
{
  auto freekBytes = (FILESYSTEM_TO_USE.totalBytes() - FILESYSTEM_TO_USE.usedBytes()) / 1024;

  LOG_info
      << millis() << " ms: Checking free space limit (" << MIN_FREE_KILOBYTES_ON_FILESYSTEM
      << " kB) for the logging data on the filesystem:\n"
      << " \t Available " << freekBytes
      << " kB from total " << (FILESYSTEM_TO_USE.totalBytes() / 1024) << " kB \n";

  if (freekBytes > MIN_FREE_KILOBYTES_ON_FILESYSTEM)
  {
    LOG_info << " \t -> OK: Enough free space" << std::endl;
  }
  else
  {
    LOG_info << " \t -> Min. free space limit was undercut -> Oldest logfile in directory "
             << DATA_DIRECTORY << " will be deleted ..." << std::endl;

    auto rootDir = FILESYSTEM_TO_USE.open(DATA_DIRECTORY, FILE_READ);
    if (rootDir)
    {
      std::list<std::string> filesList;
      File file = rootDir.openNextFile();
      while (file)
      {
        // LOG_info << "file " << file.path() << std::endl;
        if (!file.isDirectory())
          filesList.push_back(file.path());

        file = rootDir.openNextFile();
      }

      filesList.sort();
      LOG_info << "Sorted files list: \n";
      for (auto el : filesList)
        LOG_info << el << "\n";

      if (FILESYSTEM_TO_USE.remove(filesList.front().c_str()))
        LOG_info << "File   " << filesList.front() << " deleted" << std::endl;
      else
        LOG_error << millis() << " ms: ERROR: xTask_Watchdog_and_Maintenance at Check for free space on filesystem -> FILE  "
                  << filesList.front()
                  << "   CANNOT BE DELETED -> NO SPACE FREED ON FILESYSTEM (" << freekBytes << " kB available)" << std::endl;
    }
  }
}

/*********************************************************************
 * @brief 	in case the output to std::err is redirected to a syslog file
 *          -> the filebuffer is only written if file is closed (BUG?)
 *          -> for that we close and reopen the file here
 * @param 	void
 * @return 	void
 *********************************************************************/
void writeSyslogfile()
{
#if REDIRECT_STD_ERR_TO_SYSLOG_FILE
  if (syslogFileStream && syslogFileStream.is_open())
  {
    syslogFileStream.close();
    syslogFileStream.open(syslogFileName, std::ofstream::out | std::ofstream::app);
  }
#endif
}
